# Exploring Hacker News Posts Project

Hacker News is a site started by the startup incubator Y Combinator, where user-submitted stories are voted and commented upon, similar to reddit. Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of Hacker News' listings can get hundreds of thousands of visitors as a result.

The data can be accessed [here](https://www.kaggle.com/hacker-news/hacker-news-posts), but it has been reduced from almost 300,000 rows to approximately 20,000 rows by removing all the submissions that did not receive any comments, and then randomly sampling from the remaining submissions. 

We're specifically interested in posts whose titles begin with either 'Ask HN' posts, where users ask the Hacker News community a specific question, and 'Show HN' posts, where users post to show the community a project, product, or something else interesting.

We'll specifically compare these two types of posts to determine the following:

* Do Ask HN or Show HN receive more comments on average?
* Do posts created at a certain time receive more comments on average?

In [1]:
import csv

f = open('hacker_news.csv')
hn = list(csv.reader(f))
hn[:5]

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'],
 ['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01']]

The first list in the inner lists contains the column headers, and the lists after contain the data for one row. In order to analyze our data, we need to remove the row with the column headers.

In [2]:
headers = hn[0] # do not run this twice; restart and run all kernels if needed
hn = hn[1:]

print(headers)
print('\n')
hn[:5]

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']




[['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01'],
 ['10301696',
  'Note by Note: The Making of Steinway L1037 (2007)',
  'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0',
  '8',
  '2',
  'walterbell',
  '9/30/2015 4:12']]

## Extracting Ask HN and Show HN Posts

Since we're only concerned with post titles beginning with 'Ask HN' or 'Show HN', we'll create new lists of lists containing just the data for those titles. We'll be using the string method 'startswith' and the lower() method to control the case of our results and create these lists.

In [3]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1].lower()
    if title.startswith('ask hn'):
        ask_posts.append(row)
    elif title.startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)
        
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


After filtering the data, we have 1,744 'Ask' posts, 1,162 'Show' posts, and 17,194 other posts.

## Average Number of Comments for Posts

We will now be able to calculate the average number of comments for the 'Ask' and 'Show' posts and determine which typically receives more comments

In [4]:
total_ask_comments = 0

for item in ask_posts:
    total_ask_comments += int(item[4])
    
total_ask_comments / len(ask_posts)

14.038417431192661

In [5]:
total_show_comments = 0

for item in show_posts:
    total_show_comments += int(item[4])
    
total_show_comments / len(show_posts)

10.31669535283993

On average, an 'Ask' post received 14 comments, whereas a 'Show' post received 10 comments on average. Since ask posts are more likely to receive comments, we will focus our remaining analysis just on these posts.

## Finding the Amount of Ask Posts and Comments by Hour Created

We will determine if ask posts created at a certain time are more likely to attract comments. We will use the following steps to perform this analysis:

* Calculate the amount of ask posts created in each hour of the day, along with the number of comments received
* Calculate the average number of comments ask posts receive by hour created

We will use the datetime module to work with the data in the 'created_at' column.

In [6]:
import datetime as dt

result_list = []

for post in ask_posts:
    result_list.append([post[6], int(post[4])])
    
counts_by_hour = {}
comments_by_hour = {}
date_format = "%m/%d/%Y %H:%M"

for row in result_list:
    date = row[0]
    comment = row[1]
    time = dt.datetime.strptime(date, date_format).strftime("%H")
    if time in counts_by_hour:
        counts_by_hour[time] +=1
        comments_by_hour[time] += comment
    else:
        counts_by_hour[time] = 1
        comments_by_hour[time] = comment

comments_by_hour

{'00': 447,
 '01': 683,
 '02': 1381,
 '03': 421,
 '04': 337,
 '05': 464,
 '06': 397,
 '07': 267,
 '08': 492,
 '09': 251,
 '10': 793,
 '11': 641,
 '12': 687,
 '13': 1253,
 '14': 1416,
 '15': 4477,
 '16': 1814,
 '17': 1146,
 '18': 1439,
 '19': 1188,
 '20': 1722,
 '21': 1745,
 '22': 479,
 '23': 543}

## Calculate the Average Number of Comments per Hour

We will use the two dictionaries we created to calculate the average number of comments for posts created during each hour of the day.

In [7]:
avg_by_hour = []

for hr in comments_by_hour:
    avg_by_hour.append([hr, comments_by_hour[hr] / counts_by_hour[hr]])
    
avg_by_hour

[['16', 16.796296296296298],
 ['10', 13.440677966101696],
 ['21', 16.009174311926607],
 ['03', 7.796296296296297],
 ['00', 8.127272727272727],
 ['09', 5.5777777777777775],
 ['22', 6.746478873239437],
 ['14', 13.233644859813085],
 ['23', 7.985294117647059],
 ['02', 23.810344827586206],
 ['06', 9.022727272727273],
 ['20', 21.525],
 ['15', 38.5948275862069],
 ['19', 10.8],
 ['04', 7.170212765957447],
 ['12', 9.41095890410959],
 ['07', 7.852941176470588],
 ['17', 11.46],
 ['13', 14.741176470588234],
 ['11', 11.051724137931034],
 ['18', 13.20183486238532],
 ['01', 11.383333333333333],
 ['08', 10.25],
 ['05', 10.08695652173913]]

We now have the results we need, but this format makes it hard to identify the hours with the highest values. Let's finish by sorting the list of lists and printing the five highest values in a format that's easier to read.

In [8]:
swap_avg_by_hour = []

for row in avg_by_hour:
    swap_avg_by_hour.append([row[1],row[0]])

swap_avg_by_hour

[[16.796296296296298, '16'],
 [13.440677966101696, '10'],
 [16.009174311926607, '21'],
 [7.796296296296297, '03'],
 [8.127272727272727, '00'],
 [5.5777777777777775, '09'],
 [6.746478873239437, '22'],
 [13.233644859813085, '14'],
 [7.985294117647059, '23'],
 [23.810344827586206, '02'],
 [9.022727272727273, '06'],
 [21.525, '20'],
 [38.5948275862069, '15'],
 [10.8, '19'],
 [7.170212765957447, '04'],
 [9.41095890410959, '12'],
 [7.852941176470588, '07'],
 [11.46, '17'],
 [14.741176470588234, '13'],
 [11.051724137931034, '11'],
 [13.20183486238532, '18'],
 [11.383333333333333, '01'],
 [10.25, '08'],
 [10.08695652173913, '05']]

In [9]:
sorted_swap = sorted(swap_avg_by_hour, reverse=True)
sorted_swap

[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [13.20183486238532, '18'],
 [11.46, '17'],
 [11.383333333333333, '01'],
 [11.051724137931034, '11'],
 [10.8, '19'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [9.41095890410959, '12'],
 [9.022727272727273, '06'],
 [8.127272727272727, '00'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [7.170212765957447, '04'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09']]

In [10]:
print("Top 5 Hours for Ask Posts Comments")

for average, hour in sorted_swap[:5]:
    print("{}: {:.2f} average comments per post".format
          (dt.datetime.strptime(hour, '%H').strftime('%H:%M'), average)
    )

Top 5 Hours for Ask Posts Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.

According to the data set documentation, the timezone used is Eastern Time in the US. So, we could also write 15:00 as 3:00 pm est.

## Conclusion

In this project, we analyzed ask posts and show posts to determine which type of post and time receive the most comments on average. Based on our analysis, to maximize the amount of comments a post receives, we'd recommend the post be categorized as ask post and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est).

However, it should be noted that the data set we analyzed excluded posts without any comments. Given that, it's more accurate to say that of the posts that received comments, ask posts received more comments on average and ask posts created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est) received the most comments on average.